pkg/common/commands/resetoffset/controller/reconciler.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Shopify/sarama"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+
+	kafkav1alpha1 "knative.dev/eventing-kafka/pkg/apis/kafka/v1alpha1"
+	kafkasarama "knative.dev/eventing-kafka/pkg/channel/distributed/common/kafka/sarama"
+	"knative.dev/eventing-kafka/pkg/client/injection/reconciler/kafka/v1alpha1/resetoffset"
+	kafkalisters "knative.dev/eventing-kafka/pkg/client/listers/kafka/v1alpha1"
+	"knative.dev/eventing-kafka/pkg/common/client"
+	"knative.dev/eventing-kafka/pkg/common/commands/resetoffset/refmappers"
+	commonconstants "knative.dev/eventing-kafka/pkg/common/constants"
+)
+
+var (
+	_ resetoffset.Interface = (*Reconciler)(nil) // Verify Reconciler Implements Interface
+	_ resetoffset.Finalizer = (*Reconciler)(nil) // Verify Reconciler Implements Finalizer
+)
+
+// Reconciler Implements controller.Reconciler for ResetOffset Resources
+type Reconciler struct {
+	kafkaBrokers      []string
+	saramaConfig      *sarama.Config
+	resetoffsetLister kafkalisters.ResetOffsetLister
+	refMapper         refmappers.ResetOffsetRefMapper
+}
+
+// ReconcileKind implements the Reconciler Interface and is responsible for performing Offset repositioning.
+func (r *Reconciler) ReconcileKind(ctx context.Context, resetOffset *kafkav1alpha1.ResetOffset) reconciler.Event {
+
+	// Get The Logger From Context
+	logger := logging.FromContext(ctx).Desugar()
+	logger.Debug("<==========  START RESET-OFFSET RECONCILIATION  ==========>")
+
+	// Get The EventRecorder From Context
+	eventRecorder := controller.GetEventRecorder(ctx)
+
+	// Ignore Previously Initiated ResetOffset Instances
+	if resetOffset.Status.IsInitiated() || resetOffset.Status.IsCompleted() {
+		logger.Debug("Skipping reconciliation of previously executed ResetOffset instance")
+		return reconciler.NewEvent(corev1.EventTypeNormal, ResetOffsetSkipped.String(), "Skipped previously executed ResetOffset")
+	}
+
+	// Set Defaults & Verify The ResetOffset Is Valid
+	resetOffset.SetDefaults(ctx)
+	if err := resetOffset.Validate(ctx); err != nil {
+		logger.Error("Received invalid ResetOffset", zap.Error(err))
+		eventRecorder.Event(resetOffset, corev1.EventTypeWarning, ResetOffsetSkipped.String(), "Skipping invalid ResetOffset")
+		return err
```

Comment:
I agree it is redundant once we have a Webhook stood up for this, but... I was following the example in both the consolidated and distributed KafkaChannel `ReconcileKind()` implementations.  Is there a reason for re-validating in those cases?  Is it because we used to have an "open" KafkaChannel CRD that didn't explicitly specify it's payload?  Does it hurt anything to keep this here? (I'm happy to remove it - just double checking on the inconsistency ; )

Reply:
> Is there a reason for re-validating in those cases?

TBH I don't know, for sure revalidating doesn't make sense... Maybe setting defaults is important? But I don't think so too, because the webhook does also the defaulting part

Reply:
I'll add a TODO comment to that logic indicating it should be re-evaluated and likely removed once we've stood up a webhook.  Until then... if I remove it we won't have the same functionality.  I'll update the parent issue with a note on this as well.

Reply:
> I'll add a TODO comment to that logic indicating it should be re-evaluated and likely removed once we've stood up a webhook. Until then... if I remove it we won't have the same functionality. I'll update the parent issue with a note on this as well.

I am fine to have this in, temporary. Let's capture it in a GH issue

Reply:
sure - created Issue #657